File: model/anomaly_detection_util.js
class Point{
    constructor(x,y) {
        this.x = x
        this.y = y
    }
}

class Line{
    constructor(a,b) {
        this.a = a
        this.b = b
    }
    f(x){
        return this.a*x + this.b
    }
}

function avg(x){
    let sum = 0;
    for(let i = 0; i < x.length; i++ ){
        sum += x[i]; //don't forget to add the base
    }
    return sum / x.length
}

// returns the variance of X and Y
function variance(x) {
    let sum = 0;
    for (let i = 0; i < x.length; ++i) {
        sum += x[i] * x[i]
    }
    return (sum / x.length) - (avg(x) * avg(x));
}

// returns the covariance of X and Y
function cov(x, y) {
    var size = x.length
    var xy = []
    for (var i = 0; i < size; ++i) {
        xy.push(x[i] * y[i])
    }
    return avg(xy) - avg(x) * avg(y);
}

// returns the Pearson correlation coefficient of X and Y
function pearson(x, y) {
    return cov(x, y) / (Math.sqrt(variance(x)) * Math.sqrt(variance(y)));
}

// performs a linear regression and returns the line equation
function linear_reg(points) {
    var size = points.length
    let x = []
    let y = []
    for (var i = 0; i < size; ++i) {
        x.push(points[i].x);
        y.push(points[i].y);
    }
    const a = cov(x, y) / variance(x);
    const b = avg(y) - a * avg(x);
    return new Line(a, b);
}


// returns the deviation between point p and the line equation of the points
function dev(p, points) {
    return dev2(p, linear_reg(points));
}

// returns the deviation between point p and the line
function dev2(p, l) {
    return Math.abs(l.f(p.x) - p.y);
}

File: model/src/anomaly_detection_util.cpp
/* Author: 208604264 Aiden Kaminsky
 * mile stone 3 */

#include <cmath>
#include <cstdlib>
#include "anomaly_detection_util.h"

float avg(const float* x, int size){
    float sum = 0;
    for (int i = 0; i < size; i++) {
        sum += x[i];
    }
    return sum / size;
}

// returns the variance of X and Y
float var(const float* x, int size){
    float sum = 0;
    float u = avg(x, size);
    for (int i = 0; i < size; i++) {
        sum += pow(x[i], 2.0);
    }
    return (sum / size) - pow(u, 2.0);
}

// returns the covariance of X and Y
float cov(const float* x, const float* y, int size){
    float* temp = new float[size];
    for (int i = 0; i < size; i++) {
        temp[i] = x[i] * y[i];
    }
    float result = 0;
    result = avg(temp, size) - avg(x, size) * avg(y, size);
    delete[] temp;
	return result;
}

// returns the Standard deviation of Xi from 1 to size
float deviation(const float* x, int size) {
    float sum = 0;
    float ave = avg(x, size);
    for (int i = 0; i < size; i++) {
        sum += pow((x[i] - ave), 2.0);
    }
    return sqrt(sum / size);
}

// returns the Pearson correlation coefficient of X and Y
float pearson(const float* x, const float* y, int size){
	return cov(x, y, size) / (deviation(x, size) * deviation(y, size));
}

// performs a linear regression and returns the line equation
// gets: a array of points , and size
Line linear_reg(Point** points, int size){
    float* x = new float[size];
    for (int i = 0; i < size; i++) {
        x[i] = points[i]->x;
    }
    float* y = new float[size];
    for (int i = 0; i < size; i++) {
        y[i] = points[i]->y;
    }
    float a = cov(x, y, size) / var(x, size);
    float b = avg(y, size) - a * avg(x, size);
    delete[] x;
    delete[] y;
	return Line(a, b);
}

// returns the deviation between point p and the line equation of the points
float dev(Point p,Point** points, int size){
    Line l = linear_reg(points, size);
	return dev(p, l);
}

// returns the deviation between point p and the line
float dev(Point p,Line l){
    return fabs(l.f(p.x) - p.y);
}
